mne/preprocessing/asr_utils.py

```diff
@@ -0,0 +1,509 @@
+
+    # compute impulse response
+    _, Ss = 2 * np.real(signal.freqz(Qh, A, worN=n, whole=True))
+
+    hh = np.fft.ifft(
+        np.exp(np.fft.fft(Rwindow * np.fft.ifft(np.log(Ss, dtype=np.complex))))  # noqa
+    )
+    B = np.real(numf(hh[0:nr], A, nb))
+
+    return B, A
+
+
+def yulewalk_filter(X, sfreq, zi=None, ab=None, axis=-1):
+    """Yulewalk filter.
+
+    Parameters
+    ----------
+    X : array, shape = (n_channels, n_samples)
+        Data to filter.
+    sfreq : float
+        Sampling frequency.
+    zi : array, shape=(n_channels, filter_order)
+        Initial conditions.
+    a, b : 2-tuple | None
+        Coefficients of an IIR filter that is used to shape the spectrum of
+        the signal when calculating artifact statistics. The output signal
+        does not go through this filter. This is an optional way to tune the
+        sensitivity of the algorithm to each frequency component of the
+        signal. The default filter is less sensitive at alpha and beta
+        frequencies and more sensitive at delta (blinks) and gamma (muscle)
+        frequencies.
+    axis : int
+        Axis to filter on (default=-1, corresponding to samples).
+
+    Returns
+    -------
+    out : array
+        Filtered data.
+    zf :  array, shape=(n_channels, filter_order)
+        Output filter state.
+    """
+    # Set default IIR filter coefficients
+    if ab is None:
+        F = np.array([0, 2, 3, 13, 16, 40, np.minimum(
+            80.0, (sfreq / 2.0) - 1.0), sfreq / 2.0]) * 2.0 / sfreq
+        M = np.array([3, 0.75, 0.33, 0.33, 1, 1, 3, 3])
+        B, A = yulewalk(8, F, M)
+    else:
+        A, B = ab
+
+    # apply the signal shaping filter and initialize the IIR filter state
+    if zi is None:
+        out = signal.lfilter(B, A, X, axis=axis)
+        zf = None
+    else:
+        out, zf = signal.lfilter(B, A, X, zi=zi, axis=axis)
+
+    return out, zf
+
+
+def ma_filter(N, X, Zi):
+    """Run a moving average filter over the data.
+
+    Parameters
+    ----------
+    N : int
+        Length of the filter.
+    X : array, shape=(n_channels, n_samples)
+        The raw data.
+    Zi : array
+        The initial filter conditions.
+
+    Returns
+    -------
+    X : array
+        The filtered data.
+    Zf : array
+        The new fiter conditions.
+    """
+    if Zi is None:
+        Zi = np.zeros([len(X), N])
+
+    Y = np.concatenate([Zi, X], axis=1)
+    M = Y.shape[-1]
+    I_ = np.stack([np.arange(M - N),
+                   np.arange(N, M)]).astype(int)
+    S = (np.stack([-np.ones(M - N),
+                   np.ones(M - N)]) / N)
+    X = np.cumsum(np.multiply(Y[:, np.reshape(I_.T, -1)],
+                              np.reshape(S.T, [-1])), axis=-1)
+
+    X = X[:, 1::2]
+
+    Zf = np.concatenate([-(X[:, -1] * N - Y[:, -N])[:, np.newaxis],
+                         Y[:, -N + 1:]], axis=-1)
+    return X, Zf
+
+
+def geometric_median(X, tol=1e-5, max_iter=500):
```

Comment:
could this be used to do an evoked using median and not mean of epochs?

Reply:
Sure!

Reply:
should we make a more general version of this function and add it to utils?